Source/WebKit/UIProcess/PDF/WKPDFPageNumberIndicator.mm

```diff
@@ -0,0 +1,201 @@
+    [_label setAlpha:indicatorLabelOpacity];
+    [_label setAdjustsFontSizeToFitWidth:YES];
+    [_label setAutoresizingMask:UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight];
+    WebCore::PlatformCAFilters::setBlendingFiltersOnLayer([_label layer], WebCore::BlendMode::PlusDarker);
+    [[_backdropView contentView] addSubview:_label.get()];
+
+    [self updatePosition:self.frame];
+    [self setPageCount:pageCount];
+
+    return self;
+}
+
+- (void)dealloc
+{
+    [_timer invalidate];
+    [super dealloc];
+}
+
+- (void)updatePosition:(CGRect)frame
+{
+    RetainPtr webView = _webView.get();
+    if (!webView)
+        return;
+
+    UIEdgeInsets insets = UIEdgeInsetsAdd([webView _computedUnobscuredSafeAreaInset], [webView _computedObscuredInset], UIRectEdgeAll);
+    [self _moveToPoint:UIEdgeInsetsInsetRect(frame, insets).origin animated:NO completionHandler:[view = retainPtr(self)](bool originMoved) {
+        if (!originMoved)
+            return;
+        [view sizeToFit];
+        [view show];
+    }];
+}
+
+- (void)setCurrentPageNumber:(unsigned)currentPageNumber
+{
+    ASSERT(currentPageNumber);
+    bool labelTextShouldChange = _currentPageNumber != currentPageNumber;
+    if (labelTextShouldChange)
+        _currentPageNumber = currentPageNumber;
+    [self _updateLabel:labelTextShouldChange];
+}
+
+- (void)setPageCount:(unsigned)pageCount
+{
+    ASSERT(pageCount);
+    bool labelTextShouldChange = _pageCount != pageCount;
+    if (labelTextShouldChange)
+        _pageCount = pageCount;
+    [self _updateLabel:labelTextShouldChange];
+}
+
+- (void)show
+{
+    [UIView animateWithDuration:indicatorFadeInDuration.seconds() animations:[view = retainPtr(self)] {
+        [view setAlpha:1];
+    }];
+
+    if (_timer)
+        [_timer setFireDate:[NSDate dateWithTimeIntervalSinceNow:indicatorTimeout.seconds()]];
+    else
+        _timer = [NSTimer scheduledTimerWithTimeInterval:indicatorTimeout.seconds() target:self selector:@selector(hide:) userInfo:nil repeats:NO];
+}
+
+- (void)hide:(NSTimer *)timer
+{
+    [UIView animateWithDuration:indicatorFadeOutDuration.seconds() animations:[view = retainPtr(self)] {
+        [view setAlpha:0];
+    }];
+
+    [std::exchange(_timer, nil) invalidate];
+}
+
+- (void)_moveToPoint:(CGPoint)point animated:(BOOL)animated completionHandler:(WTF::CompletionHandler<void(bool)>&&)completionHandler
+{
+    point.x += indicatorMargin;
+    point.y += indicatorMargin;
+
+    auto animations = [view = retainPtr(self), point, completionHandler = WTFMove(completionHandler)] mutable {
+        CGRect frame = [view frame];
+        bool originMoved = static_cast<bool>(CGPointEqualToPoint(frame.origin, point) == NO);
+        frame.origin = point;
+        [view setFrame:frame];
+        completionHandler(originMoved);
+    };
+
+    if (animated)
+        [UIView animateWithDuration:indicatorMoveDuration.seconds() animations:makeBlockPtr(WTFMove(animations)).get()];
```

Comment:
Just to double check, we expect this completion handler to be called when the animation begins (not when it ends)?

For the latter, I think we'd need to call `+animateWithDuration:animations:completion:`

Reply:
I did indeed want the handler to run at the end of animations, not at the beginning. Good catch! I used `+animateWithDuration:animations:completion:` instead and changed up the completion handler signature a little to still make use of the `originMoved` information.